cloudformation/cw-dashboard-substack.cfn.yaml

```diff
@@ -0,0 +1,223 @@
+{#- updated dynamically #}
+{%- set coord = {'x': 0, 'y': 0} %}
+{#- this macro allows to update the coord variable in order to put a new metric graph #}
+{#- If the graph we want to add is too big, we set x to 0 and y to be just after the last graph #}
+{%- macro update_coord(dx, dy) %}
+  {%- if coord.update({'x': (coord.x+dx)}) %} {%- endif %}
+  {%- if coord.x+dx > 24 %} {#- the longest width allowed is 24 #}
+    {%- if coord.update({'x': 0, 'y': coord.y+dy}) %} {%- endif %}
+  {%- endif %}
+{%- endmacro %}
+{#- this macro allows to update the coord variable in order to put a new header section #}
+{#- This is useful both to put the new header and also to set coord to point just below #}
+{#- the added section header. Example of section header: "EBS Metrics" #}
+{%- macro update_coord_after_section(dy) %}
+  {%- if coord.update({'x': 0, 'y': coord.y+dy}) %} {%- endif %}
+{%- endmacro %}
+{%- set graph_width = 6 %}
+{%- set graph_height = 6 -%}
+Parameters:
+  {#- Cluster parameters #}
+  PclusterStackName:
+    Description: Name of the cluster to which this dashboard belongs
+    Type: String
+  {#- EC2 parameters #}
+  MasterInstanceId:
+    Description: ID of the Master instance
+    Type: AWS::EC2::Instance::Id
+  {#- EBS parameters #}
+  EBSVolumesIds:
+    Description: IDs of the EBS volumes used
+    Type: CommaDelimitedList
+  {#- RAID parameters #}
+  RAIDVolumesIds:
+    Description: Volume IDs of the resulted RAID EBS volumes
+    Type: CommaDelimitedList
+  {#- EFS parameters #}
+  EFSFileSystemId:
+    Description: ID of the EFS volume used
+    Type: String
+  {#- FSx parameters #}
+  FSXFileSystemId:
+    Description: ID of the FSx volume used
+    Type: String
+Resources:
+  HeadNodeDashboard:
+    Type: AWS::CloudWatch::Dashboard
+    Properties:
+      DashboardName: !Join
+        - '-'
+        - - !Ref 'PclusterStackName'
+          - HeadNode
+      DashboardBody: !Join
+        - ''
+        - - '{"widgets":['
+          {#- Head Node Instance Metrics #}
+          - '{"type":"text","x":{{ coord.x }},"y":{{ coord.y }},"width":24,"height":1,"properties":{"markdown":"\n#
+            Head Node Instance Metrics\n"}}'
+          {%- set ec2_metrics = [["CPUUtilization"],["NetworkPacketsIn", "NetworkPacketsOut"], ["NetworkIn", "NetworkOut"],
+                                 ["DiskReadBytes", "DiskWriteBytes"], ["DiskReadOps", "DiskWriteOps"]]
+          %}
+          {{- update_coord_after_section(1) }}
+          {%- for metrics_group in ec2_metrics %}
+          - !Sub ',{"type":"metric","x":{{ coord.x }},"y":{{ coord.y }},"width":{{ graph_width }},"height":{{ graph_height }},"properties":{"view":"timeSeries","stacked":false,"region":"${AWS::Region}","metrics":'
+            {%- for metric in metrics_group %}
+          - !Sub '{% if loop.first %}[{% else %},{% endif %}["AWS/EC2","{{ metric }}","InstanceId","${MasterInstanceId}"]{% if loop.last %}]}}{% endif %}'
```

Comment:
For inline if/else blocks I think you can simplify the code with something like:
`{{ '[' if loop.first else ',' }}`

Reply:
You're right, thanks! Also changed {% if loop.last %}]}}{% endif %} -> {{ ']}}' if loop.last else '' }}